### DISCUSSION TITLE
[NanoVDB] Is it really possible to modify the voxel values, and how?

### DISCUSSION THREAD

**Discussion by Nord3D** [AcademySoftwareFoundation/openvdb#1208] [Q&A] [answered]
Hi
In the NanoVDB FAQ, the limitations part, there is a mention of that in principle it's possible to "modify the values in a NanoVDB grid, but not its topology". But in all of the NanoVDB headers, there are only read-only methods such as `getValue`, and nothing for "set". 
So, my question is: if it's really possible to modify the NanoVDB values without touching a topology, please tell me how I could do it? 
Specifically, I need this in Houdini 18.5 OpenCL SOP, where I can get the values via read-only `cnanovdb_readaccessor`, but also want to set modified value in one voxel per thread.

**Comment by jmlait** [comment, marked as answer]
It is possible in principle as you can add your own functions in addition to the provided headers :>

Since you are looking at cnanovdb_readaccessor, note that it has acc->mNode[0..3].  If you first run a cnanovdb_readaccessor_getValue, you will have set the mNode[0] to a non-null if the leaf node exists.  If the leaf node doesn't exist, you can't write to it because you can't change topology.

Then, to write to it, you can use
```
cnanovdb_node0F *node = acc->mNode[0];
uint32_t n = cnanovdb_node0F_CoordToOffset(ijk);
node->mVoxels[n] = newvalue;
```

Now, you say you want it in the 18.5 OpenCL SOP....  This likely will "work".  But the danger is that Houdini will be unaware you made any changes, it doesn't support the Writabel flag so assumes the value is invariant.  This may cause some weird issues if you have the VDB show up elsewhere as Houdini may re-use the GPU buffer unaware it no longer reflects the VDB value, so you changes may "leak" to other operations.  Likewise, the changes won't write back to the VDB on the CPU side.  So if something forces the geometry off the GPU and back onto the CPU (such as moving the geometry through a CPU only node, which is most of them) your changes will vanish.

But if you have your own tight loop of GPU code the changes should persist, and if your output is something other than the VDB itself, it should work.

Which is a long way of saying "here be dragons".  The data structure supports changes - no internal pointers are invalidated by changing voxel values - but there is a good reason we've not made easy ways of doing this out of the box.

**Reply by Nord3D** [reply]
Thank you very much! So, now I better understand what to do. And, the most important, I know that it is possible somehow to modify the active voxels on GPU. If it does not work inside the OpenCL SOP, I can write it via HDK, I think.
